gossip/src/push_active_set.rs

```diff
@@ -0,0 +1,237 @@
+use {
+    crate::weighted_shuffle::WeightedShuffle,
+    indexmap::IndexMap,
+    rand::Rng,
+    solana_bloom::bloom::{AtomicBloom, Bloom},
+    solana_sdk::{native_token::LAMPORTS_PER_SOL, pubkey::Pubkey},
+    std::collections::HashMap,
+};
+
+const NUM_PUSH_ACTIVE_SET_ENTRIES: usize = 36;
+
+pub(crate) struct PushActiveSet([PushActiveSetEntry; NUM_PUSH_ACTIVE_SET_ENTRIES]);
+
+// Keys are gossip nodes to push messages to.
+// Values are which origins the node has pruned.
+#[derive(Default)]
+struct PushActiveSetEntry(IndexMap</*node:*/ Pubkey, /*origins:*/ AtomicBloom<Pubkey>>);
+
+impl PushActiveSet {
+    #[cfg(debug_assertions)]
+    const MIN_NUM_BLOOM_ITEMS: usize = 512;
+    #[cfg(not(debug_assertions))]
+    const MIN_NUM_BLOOM_ITEMS: usize = crate::cluster_info::CRDS_UNIQUE_PUBKEY_CAPACITY;
+
+    pub(crate) fn get<'a>(
+        &'a self,
+        pubkey: &Pubkey,    // This node.
+        origin: &'a Pubkey, // CRDS value owner.
+        // If true forces gossip push even if the node has pruned the origin.
+        should_force_push: impl FnMut(&Pubkey) -> bool + 'a,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> impl Iterator<Item = &Pubkey> + 'a {
+        let stake = stakes.get(pubkey).min(stakes.get(origin));
+        self.get_entry(stake).get(origin, should_force_push)
+    }
+
+    pub(crate) fn prune(
+        &self,
+        pubkey: &Pubkey,    // This node.
+        node: &Pubkey,      // Gossip node.
+        origins: &[Pubkey], // CRDS value owners.
+        stakes: &HashMap<Pubkey, u64>,
+    ) {
+        let stake = stakes.get(pubkey);
+        for origin in origins {
+            if origin == pubkey {
+                continue;
+            }
+            let stake = stake.min(stakes.get(origin));
+            self.get_entry(stake).prune(node, origin)
+        }
+    }
+
+    pub(crate) fn rotate<R: Rng>(
```

Comment:
"rotate" isn't an intuitive operation here in my mind. can't decide whether a rename or doc comment is the solution though :thinking:

Reply:
Rotate as in "key rotation".
It discards the oldest entry and adds a new one.
It also rotates entries: `[a, b, c, d]` becomes `[b, c, d, e]`.

Reply:
We need `IndexMap` because we want to preserve a consistent order for the nodes which are pushed to. Similar to current master code:
https://github.com/solana-labs/solana/blob/1cc8917b9/gossip/src/crds_gossip_push.rs#L219-L224

So if `fanout = 2` and the entries are `[a, b, c, d]`, we want to push to `a` and `b`.
But if `b` gets pruned we want to push to `a` and `c`.
Then when this gets rotated to `[b, c, d, e]`, we want to push to `c` and `d`.